content.js

```diff
@@ -0,0 +1,79 @@
+/*
+ * Referer Modifier: Modify the Referer header in HTTP requests
+ * Copyright (C) 2017-2020 Fiona Klute
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+"use strict";
+
+// [NUANCE]
+//  It is undefined whether static content scripts will run before
+//  or after dynamic content scripts. If the dynamic script ran first,
+//  engineConfig is prefilled here, otherwise engineInstance.setConfig()
+//  will be called later.
+var engineConfig;
+var engineInstance = new RefererModEngine(engineConfig);
+
+(function() {
+
+	const originalGetter = Reflect.getOwnPropertyDescriptor(Document.wrappedJSObject.prototype, "referrer").get;
```

Comment:
I'm wary of calling anything from the wrapped object, just in case it might have been modified already. This script should run first, but after all we only need it because of timing issues. Retrieving the getter from `Document.prototype` works just as well, do you see any reason not do use that?

I hope this is the last question, but we'll see. :sweat_smile:

Reply:
> I'm wary of calling anything from the wrapped object, just in case it might have been modified already.

The assumption is that this content script is run only at document-start, so there will be no problem.
The only way I can see this assumption being broken is when installing the extension for the first time and you have tabs already open.

> This script should run first, but after all we only need it because of timing issues.

Not just because of "timing issues". That is covered by the dynamic content script registered from background.js
The necessity is more because of the fact that `document.referrer` is not a simple static value but a getter and
1) can change on a live document, if you change things in `about:config`!
2) can be called cross-document

> Retrieving the getter from `Document.prototype` works just as well, do you see any reason not do use that?

Hook chaining.

Reply:
> can change on a live document, if you change things in `about:config`!

Under which circumstances would that be desirable? I want the `document.referrer` to be exactly the same as what was sent in the HTTP `Referer` header.

> Hook chaining.

Please elaborate, that unfortunately doesn't mean anything to me.

Reply:
> I want the `document.referrer` to be exactly the same as what was sent in the HTTP `Referer` header.

There are some edge cases that makes that impossible. By Mozilla's design, I might add. But they all hinges on *user/privileged* actions.
Explanation:
What Referrer values are generated are dependent on the values of preferences documented in https://wiki.mozilla.org/Security/Referrer *at the time they are accessed*. This means that, by design, `document.referrer` might not reflect the HTTP Referer header sent, provided that the user or some other privileged entities change `about:config` preferences after the HTTP headers are sent, but before the page script accesses `document.referrer`.
This means that our content script that depends on Firefox-provided `document.referrer` is impacted too.
BUT! As I mentioned, this can only be a result of privileged entities, so I don't think too much about breaking this behavior.
So... there, an obscure edge case.

> Please elaborate, that unfortunately doesn't mean anything to me.

Suppose the user installs two extensions that both wants to hook `document.referrer`, I think the expectation is that the hooks are chained together, not that one hook overrides another. If you use the getter retrieved from `Document.prototype`, you will miss this case, since your hook now goes directly to the *privileged* getter.
In addition, I don't think it is wise to use *privileged* getter for response to potentially hostile code. Less confused deputy problems that way. Safer to use *unprivileged* getters, if only for hygiene.

Reply:
More on hook chaining...
It's the idea behind Windows' [SetWindowsHookEx](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexw)

Quote:
```
Calling the CallNextHookEx function to chain to the next hook procedure is optional,
but it is highly recommended; otherwise, other applications that have installed hooks
will not receive hook notifications and may behave incorrectly as a result. You should
call CallNextHookEx unless you absolutely need to prevent the notification from
being seen by other applications.
```

Reply:
> In addition, I don't think it is wise to use privileged getter for response to potentially hostile code. Less confused deputy problems that way. Safer to use unprivileged getters, if only for hygiene.

That depends on in what context the getter runs: I want to be sure that a getter from some other source (worst case a page script) won't run get to run in add-on context. Both https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts and https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Language_Bindings/Components.utils.exportFunction seem to *imply* that functions retain their source context when passed around, but they're not very clear on the matter. If the page context getter definitely runs in its unprivileged context even if called from the content script via reflection it's not an issue. :slightly_smiling_face: Do you have any better documentation on the matter?

I see what you mean by "hook chaining" now. Clean interaction with other add-ons is definitely desirable, but only as long as it doesn't cause security issues.

Reply:
> Do you have any better documentation on the matter?

Sadly, no. But I can say that the xray environment is a mess in some semi-documented/undocumented way. Not all objects in the content script space is created equal.
A promient example I found is in the interaction between privileged and unprivileged `Reflect.apply` and `Array` instances with functions in them; an unprivileged `Reflect.apply` will not be able to access that unprivileged function even when being used to call on a non-xray-wrapped unprivileged function, but a privileged `Reflect.apply` will be able to.

Reading between the lines https://developer.mozilla.org/en-US/docs/Mozilla/Gecko/Script_security, you will see that JavaScript globals such as `Reflect`, `Object`, etc. are different between each contexts. This means content script's `Object` is not the same as `window.Object` and, more confusingly, `window.Object.create(...bla...)` is not the same as `window.wrappedJSObject.Object.create(...bla...)`; the former is creating an xray-wrapped *privileged* object, but the latter is creating non-xray-wrapped *unprivileged* object.

As far as I can tell, DOM objects are placed inconsistently. For some objects, `window.Foo.wrappedJSObject` is the same as `Foo.wrappedJSObject`, but for some, not. In fact, I run into this while developing that `MediaQueryList` and `EventTarget` hooks for rugk's project.

> but only as long as it doesn't cause security issues.

Don't worry about that. `Function.prototype.call.call(originalGetter, this);` (`_call.call(originalGetter, this)` is OK, but `originalGetter.call(this)` is not.

Reply:
I've been reading a lot more today and noticed this in the [Xray vision page](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Xray_vision):

> It's worth emphasizing that even if content tricks chrome into running some unexpected code, that code does not run with chrome privileges.

So now I'm confident we're good. :slightly_smiling_face: I'll look into the test issue with iframes, if you could update the HTMLDocument to remove the prototype check we should be good to merge.

Reply:
Do you mean this?
```
_toString.call(Reflect.getPrototypeOf(this)) !== "[object HTMLDocument]"
```

Reply:
The former. So the check should be just:
```js
if (Object.prototype.toString.call(this) !== "[object HTMLDocument]")
```

Because checking both is unnecessary, as you confirmed here: https://github.com/airtower-luna/referer-mod/pull/20#discussion_r612086397 :slightly_smiling_face:

Reply:
??? What about `HTMLDocument.prototype`? Ignore that case?
```
Reflect.getOwnPropertyDescriptor(win.Document.prototype, 'referrer').get.call(HTMLDocument.prototype);
```
`_toString.call(Reflect.getPrototypeOf(this)) !== "[object HTMLDocument]"` is specifically to check for this.

Reply:
```js
Object.prototype.toString.call(HTMLDocument.prototype)
```
also returns `"[object HTMLDocument]"`, so if `this` is the prototype (or rather its wrapper) the simple check still works. Or do you have some other case in mind?

Reply:
Those checks are to weed out everything that is not a document, or, in Mozilla terms, does not implement interface `Document`, which neither `HTMLDocument.prototype` nor `Document.prototype` are.
In normal Firefox, the only objects that implement this interface are `document` objects, whose prototypes are, confusingly, [`HTMLDocument`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDocument) whose prototypes are [`Document`](https://developer.mozilla.org/en-US/docs/Web/API/Document).
The only legitimate object (hance visible through X-ray vision) that `Object.prototype.toString.call(HTMLDocument.prototype) === "[object HTMLDocument]"` but does not internally implement the `Document` interface is `HTMLDocument.prototype`. So I check for it.

So, for `var x = HTMLDocument.prototype`:
```
Object.prototype.toString.call(x) === "[object HTMLDocument]"
Object.prototype.toString.call(Reflect.getPrototypeOf(this)) === "[object Document]"
```
But for `var x = document` or `var x = someIframe.contentWindow.document`:
```
Object.prototype.toString.call(x) === "[object HTMLDocument]"
Object.prototype.toString.call(Reflect.getPrototypeOf(this)) === "[object HTMLDocument]"
```

Reply:
Addendum:
Forgot to mention the case where someone does `var foo = new Document();`
I think we should also passthrough `foo.referrer` in that case.

See: https://developer.mozilla.org/en-US/docs/Web/API/Document/Document

Reply:
Okay, now I see your point, thanks. The `foo = new Document()` case should get caught by the existing checks, so `foo.referrer` will be unmodified. :slightly_smiling_face: